python/GafferSceneTest/PathFilterTest.py

```diff
@@ -202,5 +202,118 @@ def testEmptyStringMatchesNothing( self ) :
 			c["scene:path"] = IECore.InternedStringVectorData( [ "a" ] )
 			self.assertEqual( f["out"].getValue(), IECore.PathMatcher.Result.NoMatch )
 
+	def testRoots( self ) :
+
+		#  /outerGroup
+		#     /sphere
+		#     /cube
+		#     /innerGroup/
+		#         /sphere
+		#         /cube
+		#  /sphere
+
+		sphere = GafferScene.Sphere()
+		cube = GafferScene.Cube()
+
+		innerGroup = GafferScene.Group()
+		innerGroup["in"][0].setInput( sphere["out"] )
+		innerGroup["in"][1].setInput( cube["out"] )
+		innerGroup["name"].setValue( "innerGroup" )
+
+		outerGroup = GafferScene.Group()
+		outerGroup["in"][0].setInput( sphere["out"] )
+		outerGroup["in"][1].setInput( cube["out"] )
+		outerGroup["in"][2].setInput( innerGroup["out"] )
+		outerGroup["name"].setValue( "outerGroup" )
+
+		parent = GafferScene.Parent()
+		parent["in"].setInput( outerGroup["out"] )
+		parent["children"][0].setInput( sphere["out"] )
+		parent["parent"].setValue( "/" )
+
+		pathFilter = GafferScene.PathFilter()
+		pathFilter["paths"].setValue( IECore.StringVectorData( [ "/sphere" ] ) )
+
+		def assertMatchingPaths( scene, filter, expectedPaths ) :
+
+			matchingPaths = IECore.PathMatcher()
+			GafferScene.SceneAlgo.matchingPaths( filter, scene, matchingPaths )
+			self.assertEqual( set( matchingPaths.paths() ), expectedPaths )
+
+		assertMatchingPaths(
+			parent["out"], pathFilter["out"],
+			{ "/sphere" }
+		)
+
+		rootsFilter = GafferScene.PathFilter()
+		pathFilter["roots"].setInput( rootsFilter["out"] )
+
+		assertMatchingPaths(
+			parent["out"], pathFilter["out"],
+			set()
+		)
+
+		rootsFilter["paths"].setValue( IECore.StringVectorData( [ "/" ] ) )
+		assertMatchingPaths(
+			parent["out"], pathFilter["out"],
+			{ "/sphere" }
+		)
+
+		rootsFilter["paths"].setValue( IECore.StringVectorData( [ "/", "/outerGroup" ] ) )
+		assertMatchingPaths(
+			parent["out"], pathFilter["out"],
+			{ "/sphere", "/outerGroup/sphere" }
+		)
+
+		rootsFilter["paths"].setValue( IECore.StringVectorData( [ "/", "/outerGroup", "/outerGroup/innerGroup" ] ) )
+		assertMatchingPaths(
+			parent["out"], pathFilter["out"],
+			{ "/sphere", "/outerGroup/sphere", "/outerGroup/innerGroup/sphere" }
+		)
+
+		rootsFilter["paths"].setValue( IECore.StringVectorData( [ "/outerGroup", "/outerGroup/innerGroup" ] ) )
+		assertMatchingPaths(
+			parent["out"], pathFilter["out"],
+			{ "/outerGroup/sphere", "/outerGroup/innerGroup/sphere" }
+		)
+
+	def testSpecificRootsMatches( self ) :
```

Comment:
I don't think it's really relevant because `...` is an internal detail of the roots filter on one side and the path matcher on the other. But I've added a test anyway : https://github.com/GafferHQ/gaffer/pull/3692/commits/84790b5f7a61be35ab8ea1d2f3b532743cb73f36#diff-e3f0ce0ddd268b5c0214dec057e57b33R320-R374.